### PR TITLE
Changed emails to usernames

### PR DESCRIPTION
**Pull request by byarmis**
I know that there needs to be a migration somewhere, still figuring out how / where to do that :)

### CODE REVIEW

File: api/controllers/game/counter.js [hunk 1]
```diff
@@ -15,14 +15,15 @@ module.exports = function (req, res) {
...
+                logEntry = `${player.username} played the ${card.name} to counter ${opponent.username}'s ${game.twos[game.twos.length - 1].name}.`;
               } else {
-                logEntry = `${userService.truncateEmail(player.email)} played the ${card.name} to counter ${userService.truncateEmail(opponent.email)}'s ${game.oneOff.name}.`;
+                logEntry = `${player.username} played the ${card.name} to counter ${opponent.username}'s ${game.oneOff.name}.`;
               }
               const gameUpdates = {
                 lastEvent: {
                   change: 'counter',
                   pNum: req.session.pNum,
+                  log: [game.log, ...logEntry]
```
Comment (byarmis) [inline api/controllers/game/counter.js:26]: Think this was a bug

**Comment by byarmis** [comment]
The only SQL that needs to be run I think is:

```sql
ALTER TABLE user
    RENAME COLUMN email TO username;
```

File: api/policies/hasValidUsername.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * @module      :: Policy
+ * @description :: Only allows requests that contain a 'username' parameter, which is a string and does not contain spaces
+ * @docs        :: http://sailsjs.org/#!/documentation/concepts/Policies
+ *
+ */
+module.exports = function (req, res, next) {
+  if (req.body.hasOwnProperty('username')) {
+    const username = req.body.username;
+    if (typeof (username) === 'string') {
+      if (username.match(/^[0-9a-zA-Z.@]+$/)){ // . and @ are included to allow for emails
```
Comment (byarmis) [inline api/policies/hasValidUsername.js:13]: Can you run 

```
select string_agg(c,'' order by c)
from (
  select distinct regexp_split_to_table(lower(email),'') as c
  from users
) t
```

just to get a list of the unique characters in people's email addresses so they can be added here?  Just realized there might be `_`, for instance

File: api/controllers/game/counter.js (hunk 1 above)
Comment (itsalaidbacklife) [reply api/controllers/game/counter.js:26]: Good catch!

Comment (itsalaidbacklife) [reply api/controllers/game/counter.js:26]: But the spread operator is backwards. It should be:
```
log: [...game.log, logEntry]
```

`...game.log` unpacks the array into its constituent elements so generally `[...someArray]`, makes a copy of `someArray` with the exact same elements, and `[...someArray, newElement]` creates a copy of `someArray` that also has `newElement` at the end of it, which is what we want for the game log

File: api/controllers/game/draw.js [hunk 3]
```diff
@@ -32,7 +32,7 @@ module.exports = function (req, res) {
       ];
       const gameUpdates = {
         topCard: null,
-        log: [...game.log, userService.truncateEmail(user.email) + " drew a card"],
+        log: [...game.log, user.username + " drew a card"],
```
Comment (itsalaidbacklife) [inline api/controllers/game/draw.js:35]: It's only tangentially related (and not introduced by this change), but while we're here, why don't we start using ``-strings (template strings) for consistency and readibility:
```
 log: [...game.log, `${user.username} drew a card`],
```

File: api/controllers/game/face-card.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ module.exports = function (req, res) {
           if ((card.rank >= 12 && card.rank <= 13) || card.rank === 8) {
             if (player.frozenId !== card.id) {
               // Everything okay; make changes
-              let logEntry = userService.truncateEmail(player.email) + " played the " + card.name;
+              let logEntry = player.username + " played the " + card.name;
```
Comment (itsalaidbacklife) [inline api/controllers/game/face-card.js:13]: Let's do the same and use a template string for all variabalized strings:
```
let logEntry = `${player.username} played the ${card.name}`;
```

File: api/policies/hasValidUsername.js (hunk 2 above)
Comment (itsalaidbacklife) [reply api/policies/hasValidUsername.js:13]: -.@0123456789abcdefghijklmnopqrstuvwxyz

Comment (itsalaidbacklife) [reply api/policies/hasValidUsername.js:13]: What does this regex actually check?

File: api/policies/hasValidUsername.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        return next();
+      } else {
```
Comment (itsalaidbacklife) [inline api/policies/hasValidUsername.js:15]: I'd like to start switching to a one-line early-return syntax. Can we refactor this to look like

```
if (typeof (username) !== 'string') return res.badRequest({message: "Your username must contain only letters or numbers"});
// doin stuff
```

File: api/policies/hasValidUsername.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        if (username.length === 0) {
+          return res.badRequest({message: 'Please provide a valid username'});
+        }
+        return res.badRequest({message: "Your username must contain only letters or numbers"});
```
Comment (itsalaidbacklife) [inline api/policies/hasValidUsername.js:19]: If we are going to accept @'s etc we should adjust this error message accordingly. I'm not actually sure what's reasonable to require. No empty strings, idk what else tbh. Thoughts?

File: views/homepage.ejs [hunk 7]
```diff
@@ -34,8 +34,8 @@
                             <p class="col-xs-2 col-xs-offset-5">Sign Up</p>
                         </div>
                         <div class="row">
-                            <p class="col-xs-offset-2 col-xs-2">Email</p>
-                            <input ng-model="signup.email" type="text" class="col-xs-4" required>
+                            <p class="col-xs-offset-2 col-xs-2">Username</p>
```
Comment (itsalaidbacklife) [inline views/homepage.ejs:37]: Lol thank you for maintaining this but omg this is sooooo old and not used. This whole file is completely deprecated haha oops

**Review by itsalaidbacklife — CHANGES_REQUESTED** [review]
Thanks for this update! Requesting a few chores and let's discuss username validation b/c I don't really know what even makes sense as a requirement

File: api/policies/hasValidUsername.js (hunk 5 above)
Comment (itsalaidbacklife) [reply api/policies/hasValidUsername.js:15]: Not positive how to nicely format long string lines for that, is my only concern. But early returns seem nice and I'm liking the notion of one-lining them

Comment (itsalaidbacklife) [reply api/policies/hasValidUsername.js:15]: Update: I originally suggested using a one-liner syntax for the early return like
```
if (condition) return res.badRequest({message: 'some error'});
```
but this is hard to read for long errors. I edited the above to show a more conventional multiline approach; let's do that. I'd still like to ask this to be refactored to leverage early returns

File: api/controllers/game/counter.js (hunk 1 above)
Comment (byarmis) [reply api/controllers/game/counter.js:26]: dope!  yep, fixed

File: api/policies/hasValidUsername.js (hunk 2 above)
Comment (byarmis) [reply api/policies/hasValidUsername.js:13]: It checks that there's at least one valid character (alphanumeric, `.`, `-`, or `@`).  Unless you think that usernames should be any unicode?

File: api/policies/hasValidUsername.js (hunk 6 above)
Comment (byarmis) [reply api/policies/hasValidUsername.js:19]: I think we should accept @s just so people can continue to use what they had before; as for new names, I think it's ok if we don't say exactly what can be in it, just letters and numbers?

**Comment by itsalaidbacklife** [comment]
> The only SQL that needs to be run I think is:
> 
> ```sql
> ALTER TABLE user
>     RENAME COLUMN email TO username;
> ```

Just to be safe:
```
ALTER TABLE public."user"
    RENAME COLUMN email TO username;
```

**Review by itsalaidbacklife — APPROVED** [review]
Nice improvement! Hopefully this will make signing up less intimidating and generally clarify who is who

**Pull request merged**